src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,170 @@
+    }
+
+
     @Override
     public void add(T value) {
-
+        int index = cursor;
+        ArrayList.this.add(value, index);
+        cursor = index + 1;
     }
 
+
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int s = size;
+            if (s == array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index,
+                array, index + 1,
+                s - index);
+            array[index] = value;
+            size = s + 1;
+        }
     }
 
+
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            T[] tempArray = (T[]) listToArray(list);
+            int newCapacity = size + tempArray.length;
+            array = Arrays.copyOf(array, newCapacity);
+            System.arraycopy(tempArray, 0, array, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
+
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            return array[index];
+        }
     }
 
+
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            array[index] = value;
+        }
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index);
+            }
+            size = newSize;
+            array[newSize] = null;
+            return oldValue;
+        }
     }
 
+
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found:
+        {
+            if (element == null) {
+                for (; index < size; index++) {
+                    if (array[index] == null) {
+                        remove(index);
+                        break found;
+                    }
+                }
+            } else {
+                for (; index < size; index++) {
+                    if (element.equals(array[index])) {
+                        remove(index);
+                        break found;
+                    }
+                }
+                throw new NoSuchElementException();
+            }
+        }
+        return element;
     }
 
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
+
     @Override
     public boolean isEmpty() {
```

Comment:
This method checks if arrays have any null.
It doesnt check if array is empty then consider to change name of method in this meaning.
If you wanna simulate real implementation of ArrayList, isEmpty method should check its length to zero.
```
public boolean isEmpty() {
   return size == 0;
{
```